data/frame_json.go

```diff
@@ -0,0 +1,480 @@
+				stream.WriteObjectField("type")
+				stream.WriteString(t)
+				started = true
+			}
+
+			if f.Labels != nil {
+				if started {
+					stream.WriteMore()
+				}
+				stream.WriteObjectField("labels")
+				stream.WriteVal(f.Labels)
+				started = true
+			}
+
+			if f.Config != nil {
+				if started {
+					stream.WriteMore()
+				}
+				stream.WriteObjectField("config")
+				stream.WriteVal(f.Config)
+				started = true
+			}
+
+			stream.WriteObjectEnd()
+		}
+		stream.WriteArrayEnd()
+
+		stream.WriteObjectEnd()
+	}
+
+	if includeData {
+		if includeSchema {
+			stream.WriteMore()
+		}
+
+		rowCount, err := frame.RowLen()
+		if err != nil {
+			return err
+		}
+
+		stream.WriteObjectField("data")
+		stream.WriteObjectStart()
+
+		entities := make([]*fieldEntityLookup, len(frame.Fields))
+		entityCount := 0
+
+		stream.WriteObjectField("fields")
+		stream.WriteArrayStart()
+		for fidx, f := range frame.Fields {
+			if fidx > 0 {
+				stream.WriteMore()
+			}
+			isTime := f.Type().Time()
+
+			stream.WriteArrayStart()
+			for i := 0; i < rowCount; i++ {
+				if i > 0 {
+					stream.WriteRaw(",")
+				}
+				v, ok := f.ConcreteAt(i)
+				if ok {
+					if isTime {
+						v = v.(time.Time).UnixNano() / int64(time.Millisecond) // Millisconds precision
+					}
+
+					stream.WriteVal(v)
+					if stream.Error != nil { // NaN +Inf/-Inf
+						txt := fmt.Sprintf("%v", v)
+						if entities[fidx] == nil {
+							entities[fidx] = &fieldEntityLookup{}
+						}
+						entities[fidx].add(txt, i)
+						entityCount++
+						stream.Error = nil
+						stream.WriteNil()
+					}
+				} else {
+					stream.WriteNil()
+				}
+			}
+			stream.WriteArrayEnd()
+		}
+		stream.WriteArrayEnd()
+
+		if entityCount > 0 {
+			if started {
+				stream.WriteMore()
+			}
+			stream.WriteObjectField("entities")
+			stream.WriteVal(entities)
+		}
+
+		stream.WriteObjectEnd()
+	}
+	stream.WriteObjectEnd()
+	return nil
+}
+
+// ArrowToJSON writes a frame to JSON
+func ArrowBufferToJSON(b []byte) ([]byte, error) {
+	fB := filebuffer.New(b)
+	fR, err := ipc.NewFileReader(fB)
+	if err != nil {
+		return nil, err
+	}
+	defer fR.Close()
+
+	record, err := fR.Read()
+	if errors.Is(err, io.EOF) {
+		return nil, fmt.Errorf("No records found")
+	}
+	if err != nil {
+		return nil, err
+	}
+	// TODO?? multiple records in one file?
+
+	return ArrowToJSON(record, true, true)
+}
+
+// ArrowToJSON writes a frame to JSON
+func ArrowToJSON(record array.Record, includeSchema bool, includeData bool) ([]byte, error) {
+	cfg := jsoniter.ConfigCompatibleWithStandardLibrary
+	stream := cfg.BorrowStream(nil)
+	defer cfg.ReturnStream(stream)
+
+	err := writeArrowFrame(stream, record, includeSchema, includeData)
+	if err != nil {
+		return nil, err
+	}
+
+	if stream.Error != nil {
+		fmt.Println("error:", stream.Error)
+		return nil, stream.Error
+	}
+	return stream.Buffer(), nil
+}
+
+func writeArrowFrame(stream *jsoniter.Stream, record array.Record, includeSchema bool, includeData bool) error {
```

Comment:
any suggestions for this function?  
![image](https://user-images.githubusercontent.com/705951/107844466-74f31680-6d88-11eb-89c2-3c36b7cf759d.png)

Reply:
Think you can ignore them as a start. But it basically means the code is too complex :smile: Maybe extracting includeSchema and includeData in two separate functions would resolve it?

```suggestion
//gocyclo:ignore
func writeArrowFrame(stream *jsoniter.Stream, record array.Record, includeSchema bool, includeData bool) error {
```